File: play-music.c
#define _POSIX_C_SOURCE 199309L
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
// POSIX.
#include <dirent.h>
#include <regex.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>
// From include/.
#include <anal.h>

////////////////////////////////////////////////////////////////////////////////
// Utilities                                                                  //
////////////////////////////////////////////////////////////////////////////////

#define ARRAY_SIZE(array) (sizeof(array)/sizeof((array)[0]))

// Logging message prefixes.
#define INFO  "INFO: "
#define WARN  "WARN: "
#define ERROR "ERROR:"

// Deinitialize with free().
NONNULL static char* cstrCopy(const char *const cstr) {
    assert(cstr);

    char *const result = calloc(1 + strlen(cstr), 1);
    if (NULL == result) {
        perror(ERROR" Failed to allocate memory. Buy more RAM lol");
        exit(1);
    }
    return strcpy(result, cstr);
}

// TODO: check exit status.
NONNULL static void runCommand(
    const char *const        program,
    const char *const *const arguments,
    const size_t             arguments_length
) {
    assert(program);
    assert(arguments);
#ifndef NDEBUG
    for (size_t i = 0; i < arguments_length; ++i) assert(arguments[i]);
#endif

    const pid_t pid = fork();
    if (-1 == pid) {
        perror(ERROR" Failed to spawn child process");
        exit(1);
    }

    if (0 != pid) {
        wait(NULL);
        return;
    }

    // int execvp(const char *file, char *const argv[]);
    //                              ^
    // Who the FUCK wrote this type signature?

    // Create argv.
    const size_t argc = 1 + arguments_length;
    char*        argv[1 + argc];
    argv[0] = cstrCopy(program); // Must free().
    for (size_t i = 0; i < arguments_length; ++i) {
        argv[1 + i] = cstrCopy(arguments[i]); // Must free().
    }
    argv[argc] = NULL;

    // Display program to run.
    printf(INFO" Running command '%s", argv[0]);
    for (size_t i = 1; i < argc; ++i) {
        printf(" %s", argv[i]);
    }
    printf("'\n");

    // Finally run program. Sheeeeesh.
    if (-1 == execvp(program, argv)) {
        perror(ERROR" Failed to run command");
        exit(1);
    }

    for (size_t i = 0; i < argc; ++i) {
        free(argv[i]);
        argv[i] = NULL;
    };
}

static const char *const music_file_extensions[] = {
    ".mp3",
    ".flac",
    ".wav",
    ".ogg"
};

NONNULL static bool isMusicFile(const char *const path) {
    assert(path);

    const size_t path_length = strlen(path);

    const char* file_extension  = path + path_length - 1;
    size_t      characters_left = path_length;
    while (0 < characters_left) {
        if ('.' == *file_extension || '/' == *file_extension) break;
        --file_extension;
        --characters_left;
    }

    for (size_t i = 0; i < ARRAY_SIZE(music_file_extensions); ++i) {
        if (0 == strcmp(file_extension, music_file_extensions[i])) {
            return true;
        }
    }
    return false;
}

////////////////////////////////////////////////////////////////////////////////
// Playlists                                                                  //
////////////////////////////////////////////////////////////////////////////////

// Zero initialized.
// Deinitialize with playlist_deinit().
typedef struct {
    size_t size;
    size_t count;
    char** songs; // File paths.
} Playlist;

static const size_t playlist_initial_size = 50;

NONNULL static void playlistDeinit(Playlist *const playlist) {
    assert(playlist);

    for (size_t i = 0; i < playlist->count; ++i) free(playlist->songs[i]);
    free(playlist->songs);
}

// Takes ownership of passed in path.
NONNULL static void playlistAppendOwnedSong(
    Playlist *const playlist,
    char *const     path
) {
    assert(playlist);
    assert(path);

    if (playlist->count == playlist->size) {
        if (0 == playlist->size) {
            playlist->size = playlist_initial_size;
        } else {
            playlist->size *= 2;
        }
        playlist->songs =
            realloc(playlist->songs, playlist->size * sizeof(const char*));
    }

    playlist->songs[playlist->count] = path;
    ++playlist->count;
}

// If match is not NULL, only songs whose filename matches the regex will be
// added to the playlist.
// Returns the number of songs appended.
NONNULL_ARGUMENTS(1,2) static size_t playlistAppendFromDirectory(
    Playlist *const   playlist,
    const char *const path,
    regex_t *const    match
) {
    assert(playlist);
    assert(path);

    size_t songs_appended = 0;

    DIR* dir = opendir(path); // Must closedir();
    if (NULL == dir) {
        perror(ERROR" Unable to open directory");
        exit(1);
    }
    while (true) {
        const struct dirent *const entry = readdir(dir);
        if (NULL == entry) break;

        if (!isMusicFile(entry->d_name)) continue;

        if (match && 0 != regexec(match, entry->d_name, 0, NULL, 0)) {
            continue;
        }

        char* file = // Must free(). 1 for null terminator, 1 for '/'.
            calloc(1 + 1 + strlen(path) + strlen(entry->d_name), 1);
        strcat(strcat(strcat(file, path), "/"), entry->d_name);
        playlistAppendOwnedSong(playlist, file);

        ++songs_appended;
    }
    closedir(dir);
    dir = NULL;

    return songs_appended;
}

NONNULL static void playlistShuffle(Playlist *const playlist) {
    assert(playlist);

    // https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
    for (size_t i = 0; i < playlist->count; ++i) {
        const size_t j     = i + (size_t)rand() % (playlist->count - i);
        char *const song   = playlist->songs[i];
        playlist->songs[i] = playlist->songs[j];
        playlist->songs[j] = song;
    }
}

////////////////////////////////////////////////////////////////////////////////
// CLI                                                                        //
////////////////////////////////////////////////////////////////////////////////

// Zero intialized.
typedef struct {
    size_t              count;
    const char *const * data;
} CstrSlice;

// NULL if the cstr slice is empty.
NONNULL static const char* cstrSliceHead(CstrSlice *const slice) {
    assert(slice);

    if (0 == slice->count) return NULL;

    const char *const result = slice->data[0];
    --slice->count;
    ++slice->data;
    return result;
}

#define PARSED_ARGUMENTS_MAX_DIRECTORIES 50

// Zero initialized.
typedef struct {
    const char* program_name;
    bool        dont_shuffle;
    const char* match; // NULL for no matching.
    bool        dont_repeat;

    size_t      directory_count;
    const char* directories[PARSED_ARGUMENTS_MAX_DIRECTORIES];
} ParsedArguments;

NONNULL static void parsedArgumentsAppendDirectory(
    ParsedArguments *const parsed_arguments,
    const char *const      directory
) {
    assert(parsed_arguments);
    assert(directory);

    assert(parsed_arguments->directory_count < PARSED_ARGUMENTS_MAX_DIRECTORIES);
    parsed_arguments->directories[parsed_arguments->directory_count] = directory;
    ++parsed_arguments->directory_count;
}

NONNULL static void printHelp(const char *const program) {
    assert(program);

    printf(
        "Usages:\n"
        "  %s [OPTION...] [--] DIRECTORY...\n"
        "\n"
        "Plays the music files located in DIRECTORY with mpv.\n"
        "\n"
        "Options:\n"
        "  -h, --help    Display help and exit.\n"
        "\n"
        "  -m, --match REGEX\n"
        "    Only plays songs whose file name matches REGEX.\n"
        "    REGEX is interpreted as an extended regular expression (see\n"
        "    regex(3).)\n"
        "\n"
        "  --no-shuffle\n"
        "    Plays the songs in the order they appear in the directory\n"
        "    instead of randomly shuffling them.\n"
        "\n"
        "  --no-repeat\n"
        "    Exits once all the songs have been played instead of repeating\n"
        "    them in an endless loop.\n",
        program
    );
}

NONNULL static void printShortHelp(FILE *const to, const char *const program) {
    assert(to);
    assert(program);

    fprintf(to, "Try '%s -h' for more information\n", program);
}

// Zero initialized.
typedef enum {
    ARGUMENT_PARSER_BASE = 0,
    ARGUMENT_PARSER_END_OF_OPTIONS,
    ARGUMENT_PARSER_MATCH
} ArgumentParserState;

// options is the list of short options without the preceeding '-'.
NONNULL static void parseShortOptions(
    CstrSlice *const       arguments,
    ParsedArguments *const parsed_arguments,
    const char *const      options
) {
    assert(arguments);
    assert(parsed_arguments);
    assert(options);

    const char* next = options;
    while ('\0' != *next) {
        switch (*next) {
        case 'h': {
            printHelp(parsed_arguments->program_name);
            exit(0);
        }

        case 'm': {
            ++next;
            // If there is leftover options, it is the argument to -m.
            if ('\0' != *next) {
                parsed_arguments->match = next;
                return;
            }
            // Else, the next argument is.
            const char *const next_argument = cstrSliceHead(arguments);
            printf("got: %s\n", next_argument);
            if (NULL == next_argument) {
                fprintf(
                    stderr,
                    ERROR" Option '-m' expects a regular expression as an argument\n"
                );
                printShortHelp(stderr, parsed_arguments->program_name);
                exit(1);
            }
            parsed_arguments->match = next_argument;
            return;
        }

        default: {
            fprintf(stderr, ERROR" Unknown option '-%c'\n", *next);
            printShortHelp(stderr, parsed_arguments->program_name);
            exit(1);
        }
        }

        ++next;
    }
}

NONNULL static ParsedArguments parseArguments(
    const int                argc,
    const char *const *const argv
) {
    assert(argv);
#ifndef NDEBUG
    for (int i = 0; i < argc; ++i) assert(argv[i]);
#endif

    CstrSlice arguments = {
        .count = (size_t)argc,
        .data  = argv
    };
    ParsedArguments     parsed_arguments = {0};
    ArgumentParserState state            = {0};

    parsed_arguments.program_name = cstrSliceHead(&arguments);

    while (true) {
        switch (state) {
        case ARGUMENT_PARSER_BASE: {
            const char *const next = cstrSliceHead(&arguments);
            if (NULL == next)      goto largument_parser_end;
            if (0 == strlen(next)) break;

            if (0 == strcmp("--help", next)) {
                printHelp(parsed_arguments.program_name);
                exit(0);
            }
            if (0 == strcmp("--no-shuffle", next)) {
                parsed_arguments.dont_shuffle = true;
                break;
            }
            if (0 == strcmp("--match", next)) {
                state = ARGUMENT_PARSER_MATCH;
                break;
            }
            if (0 == strcmp("--no-repeat", next)) {
                parsed_arguments.dont_repeat = true;
                break;
            }
            if (0 == strcmp("--", next)) {
                state = ARGUMENT_PARSER_END_OF_OPTIONS;
                break;
            }
            if (0 == strncmp("--", next, 2)) {
                fprintf(stderr, ERROR" Unknown option '%s'\n", next);
                exit(1);
            }
            if ('-' == next[0]) {
                // + 1 to skip '-'.
                parseShortOptions(&arguments, &parsed_arguments, 1 + next);
                break;
            }

            parsedArgumentsAppendDirectory(&parsed_arguments, next);
            break;
        }

        case ARGUMENT_PARSER_END_OF_OPTIONS: {
            const char *const next = cstrSliceHead(&arguments);
            if (NULL == next) goto largument_parser_end;
            parsedArgumentsAppendDirectory(&parsed_arguments, next);
            break;
        }

        case ARGUMENT_PARSER_MATCH: {
            const char *const next = cstrSliceHead(&arguments);
            if (NULL == next) {
                fprintf(
                    stderr,
                    ERROR" Option '--match' expects a regular expression as an argument\n"
                );
                printShortHelp(stderr, parsed_arguments.program_name);
                exit(1);
            }
            parsed_arguments.match = next;
            state = ARGUMENT_PARSER_BASE;
            break;
        }

        default: {
            assert(false && "unreachable");
            exit(1);
        }
        }
    }
largument_parser_end:

    return parsed_arguments;
}

// TODO: add configuration file.
// TODO: add way to specify arguments for mpv?
// TODO: add way to override supported file extensions.

int main(const int argc, const char *const *const argv) {
    // Intialize random number generator.
    struct timespec time;
    if (-1 == clock_gettime(CLOCK_MONOTONIC, &time)) {
        perror(ERROR" Failed to read from monotonic clock");
        exit(1);
    }
    srand((unsigned int)time.tv_sec);

    const ParsedArguments arguments = parseArguments(argc, argv);

    if (0 == arguments.directory_count) {
        fprintf(stderr, ERROR" No directories specified\n");
        printShortHelp(stderr, arguments.program_name);
        exit(1);
    }

    regex_t regex;
    if (arguments.match) {
        const int result = // Must regfree().
            regcomp(&regex, arguments.match, REG_EXTENDED | REG_ICASE);

        if (0 != result) {
            const size_t error_size = regerror(result, &regex, NULL, 0);
            char error[error_size];
            regerror(result, &regex, error, error_size);
            fprintf(
                stderr,
                ERROR" Failed to compile match expression '%s': %s\n",
                arguments.match, error
            );
            exit(1);
        }
    }

    Playlist playlist = {0};

    for (size_t i = 0; i < arguments.directory_count; ++i) {
        const char *const directory = arguments.directories[i];
        printf(INFO" Loading music from directory '%s'...\n", directory);
        const size_t songs_loaded =
            playlistAppendFromDirectory(
                &playlist,
                directory,
                arguments.match ? &regex : NULL
            );

        if (0 == songs_loaded) {
            fprintf(stderr, WARN" Directory empty. Skipping...\n");
            continue;
        }
    }

    if (arguments.match) regfree(&regex);

    if (!arguments.dont_shuffle) playlistShuffle(&playlist);

    if (0 == playlist.count) {
        fprintf(stderr, ERROR" No songs loaded\n");
        exit(1);
    } else {
        printf(INFO" %zu songs loaded\n", playlist.count);
    }

    do {
        for (size_t song = 0; song < playlist.count; ++song) {
            static const char* args[1];
            args[0] = playlist.songs[song];
            runCommand("mpv", args, ARRAY_SIZE(args));
        }
    } while (!arguments.dont_repeat);

    playlistDeinit(&playlist);

    return 0;
}

File: play-music.zig
//! This file is part of play-music.
//!
//! Copyright (c) 2025 ona-li-toki-e-jan-Epiphany-tawa-mi
//!
//! play-music is free software: you can redistribute it and/or modify it under
//! the terms of the GNU General Public License as published by the Free
//! Software Foundation, either version 3 of the License, or (at your option)
//! any later version.
//!
//! play-music is distributed in the hope that it will be useful, but WITHOUT
//! ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//! FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
//! more details.
//!
//! You should have received a copy of the GNU General Public License along with
//! play-music. If not, see <https://www.gnu.org/licenses/>.

const std = @import("std");
const debug = std.debug;
const fs = std.fs;
const io = std.io;
const mem = std.mem;
const process = std.process;
const time = std.time;

const Allocator = std.mem.Allocator;
const ArgIterator = std.process.ArgIterator;
const ArrayListUnmanaged = std.ArrayListUnmanaged;
const AutoHashMapUnmanaged = std.hash_map.AutoHashMapUnmanaged;
const BufferedWriter = std.io.BufferedWriter;
const Child = std.process.Child;
const File = std.fs.File;
const GeneralPurposeAllocator = std.heap.GeneralPurposeAllocator;
const Random = std.Random;
const StaticStringMap = std.StaticStringMap;

const BufferedFileWriter = BufferedWriter(4096, File.Writer);
const RandomPrng = Random.DefaultPrng;

fn bufferedFileWriter(writer: File.Writer) BufferedFileWriter {
    return .{ .unbuffered_writer = writer };
}

////////////////////////////////////////////////////////////////////////////////
// CLI                                                                        //
////////////////////////////////////////////////////////////////////////////////

pub fn main() !void {
    var stderr = bufferedFileWriter(io.getStdErr().writer());
    defer stderr.flush() catch {};
    var stdout = bufferedFileWriter(io.getStdOut().writer());
    defer stdout.flush() catch {};

    var gpa = GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    defer _ = gpa.deinit();

    ParsedArguments.init(
        allocator,
        &stderr,
        &stdout,
    ) catch |err| switch (err) {
        error.ExitSuccess => return,
        else => return err,
    };
    defer ParsedArguments.deinit();

    var prng = RandomPrng.init(@as(u64, @bitCast(time.milliTimestamp())));
    const random = prng.random();

    try SoundSystem.init(allocator);
    defer SoundSystem.deinit();

    var playlist = Playlist.init(allocator);
    defer playlist.deinit();
    for (ParsedArguments.directories.items) |directory| {
        const songs_loaded = try playlist.appendFromDirectory(&stderr, directory);
        try stdout.writer().print(
            "INFO: {d} song(s) loaded from directory: {s}\n",
            .{ songs_loaded, directory },
        );
    }
    if (ParsedArguments.shuffle) playlist.shuffle(random);

    {
        const songs_loaded = playlist.songs.items.len;
        if (0 == songs_loaded) {
            try stderr.writer().print("ERROR: No songs were found\n", .{});
            return error.NoSongsLoaded;
        }
        try stdout.writer().print(
            "INFO: {d} song(s) loaded in total\n",
            .{songs_loaded},
        );
    }

    while (true) {
        for (playlist.songs.items) |song| {
            try stdout.writer().print("INFO: Now playing: {s}\n", .{song.file_path});
            try stderr.flush();
            try stdout.flush();
            try SoundSystem.playSong(song);
        }

        if (!ParsedArguments.repeat) break;
    }
}

fn printHelp(to: *BufferedFileWriter) !void {
    try to.writer().print(
        \\Usages:
        \\  {s} [OPTION...] [--] DIRECTORY...
        \\
        \\Plays the music files located in DIRECTORY.
        \\
        \\ Available play strategies (in order of priority):
        \\   1. With mpv if present.
        \\   2. With cvlc if present.
        \\
        \\Options:
        \\  -h, --help    Display help and exit.
        \\
        \\  -m, --match REGEX
        \\    Only plays songs whose file name matches REGEX.
        \\    REGEX is interpreted as an extended regular expression (see
        \\    regex(3).)
        \\
        \\  --no-shuffle
        \\    Plays the songs in the order they appear in the directory,
        \\    instead of randomly shuffling them.
        \\
        \\  --no-repeat
        \\    Exits once all the songs have been played, instead of repeating
        \\    them in an endless loop.
        \\
        \\  --no-skip-unplayable
        \\    Exits if some of the songs cannot be played, instead of skipping
        \\    them.
    , .{ParsedArguments.program_name});
}

fn printShortHelp(to: *BufferedFileWriter) !void {
    try to.writer().print(
        "Try '{s} -h' for more information\n",
        .{ParsedArguments.program_name},
    );
}

const ParsedArguments = struct {
    var initialized = false;
    var allocator: Allocator = undefined;

    var program_name: []const u8 = undefined;
    var directories: ArrayListUnmanaged([]u8) = undefined;
    var match: ?[]u8 = undefined;
    var shuffle: bool = undefined;
    var repeat: bool = undefined;
    var skip_unplayable: bool = undefined;

    /// Deinitialize with `deinit`.
    fn init(
        allocatorr: Allocator,
        stderr: *BufferedFileWriter,
        stdout: *BufferedFileWriter,
    ) !void {
        debug.assert(!initialized);

        allocator = allocatorr;
        directories = ArrayListUnmanaged([]u8).empty;
        match = null;
        shuffle = true;
        repeat = true;
        skip_unplayable = true;
        errdefer deinit();

        initialized = true;
        try parseArguments(stderr, stdout);
    }

    fn deinit() void {
        debug.assert(initialized);

        for (directories.items) |directory| allocator.free(directory);
        directories.deinit(allocator);
        if (match) |pattern| allocator.free(pattern);

        initialized = false;
    }

    fn appendDirectory(path: []const u8) !void {
        debug.assert(initialized);

        const path_copy = try allocator.dupe(u8, path);
        errdefer allocator.free(path_copy);
        try directories.append(allocator, path_copy);
    }

    fn setMatch(pattern: []const u8) !void {
        debug.assert(initialized);

        const pattern_copy = try allocator.dupe(u8, pattern);
        if (match) |old_pattern| allocator.free(old_pattern);
        match = pattern_copy;
    }

    fn parseArguments(
        stderr: *BufferedFileWriter,
        stdout: *BufferedFileWriter,
    ) !void {
        debug.assert(initialized);

        var arguments = try process.argsWithAllocator(allocator);
        defer arguments.deinit();

        program_name = arguments.next().?;

        while (arguments.next()) |argument| {
            if (0 == argument.len) continue;

            if (mem.eql(u8, argument, "--help")) {
                try printHelp(stdout);
                return error.ExitSuccess;
            } else if (mem.eql(u8, argument, "--no-shuffle")) {
                shuffle = false;
            } else if (mem.eql(u8, argument, "--match")) {
                if (arguments.next()) |pattern| {
                    try setMatch(pattern);
                } else {
                    try stderr.writer().print(
                        "ERROR: Option '{s}' expects a regular expression as an argument\n",
                        .{argument},
                    );
                    try printShortHelp(stderr);
                    return error.MissingMatchMattern;
                }
            } else if (mem.eql(u8, argument, "--no-repeat")) {
                repeat = false;
            } else if (mem.eql(u8, argument, "--no-skip-unplayable")) {
                skip_unplayable = false;
            } else if (mem.eql(u8, argument, "--")) {
                while (arguments.next()) |next_argument| {
                    try appendDirectory(next_argument);
                }
                return;
            } else if (2 < argument.len and mem.eql(u8, argument[0..2], "--")) {
                try stderr.writer().print(
                    "ERROR: Unknown long option '{s}'\n",
                    .{argument},
                );
                try printShortHelp(stderr);
                return error.UnknownLongOption;
            } else if ('-' == argument[0]) {
                try parseShortOptions(stderr, stdout, argument[1..], &arguments);
            } else {
                try appendDirectory(argument);
            }
        }
    }

    fn parseShortOptions(
        stderr: *BufferedFileWriter,
        stdout: *BufferedFileWriter,
        options: []const u8, // Without the trailing `-`.
        remaining_arguments: *ArgIterator,
    ) !void {
        debug.assert(initialized);

        for (0..options.len) |i| {
            const option = options[i];

            switch (option) {
                'h' => {
                    try printHelp(stdout);
                    return error.ExitSuccess;
                },
                'm' => {
                    if (i < options.len - 1) {
                        // If leftover text in options, it is the argument to -m.
                        try setMatch(options[i + 1 ..]);
                    } else if (remaining_arguments.next()) |pattern| {
                        try setMatch(pattern);
                    } else {
                        try stderr.writer().print(
                            "ERROR: Option '-{c}' expects a regular expression as an argument\n",
                            .{option},
                        );
                        try printShortHelp(stderr);
                        return error.MissingMatchMattern;
                    }
                    return;
                },
                else => {
                    try stderr.writer().print(
                        "ERROR: Unknown short option '-{c}'\n",
                        .{option},
                    );
                    try printShortHelp(stderr);
                    return error.UnknownShortOption;
                },
            }
        }
    }
};

////////////////////////////////////////////////////////////////////////////////
// Songs and Playlists                                                        //
////////////////////////////////////////////////////////////////////////////////

const FileFormat = enum {
    const Self = @This();

    flac,
    mp3,
    vorbis,
    wav,

    const extensions_formats_map = StaticStringMap(FileFormat).initComptime(.{
        .{ ".flac", .flac },
        .{ ".mp3", .mp3 },
        .{ ".ogg", .vorbis },
        .{ ".wav", .wav },
    });

    // TODO: mimetypes?
    fn formatFromFile(path: []const u8) ?Self {
        return extensions_formats_map.get(fs.path.extension(path));
    }
};

const Song = struct {
    const Self = @This();

    file_path: []u8,
    format: FileFormat,

    /// Takes ownership of passed in `path`.
    fn initFromOwnedPath(path: []u8) !Self {
        if (FileFormat.formatFromFile(path)) |format| {
            return .{
                .file_path = path,
                .format = format,
            };
        } else {
            return error.NotAnAudioFile;
        }
    }

    fn deinit(self: *Self, allocator: Allocator) void {
        allocator.free(self.file_path);
        self.* = undefined;
    }
};

const Playlist = struct {
    const Self = @This();

    allocator: Allocator,
    songs: ArrayListUnmanaged(Song),

    /// Deinitialize with `deinit`.
    fn init(allocator: Allocator) Self {
        return .{
            .allocator = allocator,
            .songs = ArrayListUnmanaged(Song).empty,
        };
    }

    fn deinit(self: *Self) void {
        for (self.songs.items) |*song| song.deinit(self.allocator);
        self.songs.deinit(self.allocator);
        self.* = undefined;
    }

    // TODO: add regex matching.
    /// Requires the sound system to be initialized (see `SoundSystem`.)
    fn appendFromDirectory(
        self: *Self,
        stderr: *BufferedFileWriter,
        path: []const u8,
    ) !u64 {
        var songs_appended: u64 = 0;

        var directory = try fs.cwd().openDir(path, .{
            .iterate = true,
        });
        defer directory.close();

        var iterator = directory.iterate();
        while (try iterator.next()) |entry| {
            var song = blk: {
                const file = try fs.path.join(
                    self.allocator,
                    &.{ path, entry.name },
                );
                errdefer self.allocator.free(file);
                break :blk Song.initFromOwnedPath(file) catch |err| switch (err) {
                    error.NotAnAudioFile => continue,
                    else => return err,
                };
            };
            errdefer song.deinit(self.allocator);

            if (!SoundSystem.isPlayable(song.format)) {
                if (ParsedArguments.skip_unplayable) {
                    try stderr.writer().print(
                        "WARN: No available strategy to play {s} files. Skipping: {s}\n",
                        .{ @tagName(song.format), song.file_path },
                    );
                } else {
                    try stderr.writer().print(
                        "ERROR: No available strategy to play {s} files. Offending file: {s}\n",
                        .{ @tagName(song.format), song.file_path },
                    );
                    return error.UnplayableFormat;
                }
            }

            try self.songs.append(self.allocator, song);
            songs_appended +|= 1;
        }

        return songs_appended;
    }

    fn shuffle(self: *Self, random: Random) void {
        random.shuffle(Song, self.songs.items);
    }
};

////////////////////////////////////////////////////////////////////////////////
// Sound System                                                               //
////////////////////////////////////////////////////////////////////////////////

// TODO: develop ability to load audio from files and pass to system's sound servers/other.

const SoundSystem = struct {
    var initialized = false;
    var allocator: Allocator = undefined;

    var formats_play_strategies_map: AutoHashMapUnmanaged(
        FileFormat,
        PlayStrategy,
    ) = undefined;

    /// Deinitialize with `deinit`.
    fn init(allocatorr: Allocator) !void {
        debug.assert(!initialized);

        allocator = allocatorr;
        formats_play_strategies_map = AutoHashMapUnmanaged(
            FileFormat,
            PlayStrategy,
        ).empty;
        errdefer formats_play_strategies_map.deinit(allocator);

        if (isProgramAvailable(allocator, &.{"mpv"})) {
            try formats_play_strategies_map.put(allocator, .flac, mpvPlayStrategy);
            try formats_play_strategies_map.put(allocator, .mp3, mpvPlayStrategy);
            try formats_play_strategies_map.put(allocator, .vorbis, mpvPlayStrategy);
            try formats_play_strategies_map.put(allocator, .wav, mpvPlayStrategy);
        } else if (isProgramAvailable(allocator, &.{ "cvlc", "-h" })) {
            try formats_play_strategies_map.put(allocator, .flac, cvlcPlayStrategy);
            try formats_play_strategies_map.put(allocator, .mp3, cvlcPlayStrategy);
            try formats_play_strategies_map.put(allocator, .vorbis, cvlcPlayStrategy);
            try formats_play_strategies_map.put(allocator, .wav, cvlcPlayStrategy);
        }

        initialized = true;
    }

    fn deinit() void {
        debug.assert(initialized);

        formats_play_strategies_map.deinit(allocator);

        initialized = false;
    }

    fn isPlayable(format: FileFormat) bool {
        debug.assert(initialized);

        return formats_play_strategies_map.contains(format);
    }

    fn playSong(song: Song) !void {
        debug.assert(initialized);

        if (formats_play_strategies_map.get(song.format)) |strategy| {
            try strategy(allocator, song);
        } else {
            return error.UnplayableFormat;
        }
    }
};

// TODO: find a better way to determine if a program is available.
/// Must have program as index 0 in `arguments`, and program should exit
/// immediately with supplied `arguments`.
fn isProgramAvailable(
    allocator: Allocator,
    arguments: []const []const u8,
) bool {
    var child = Child.init(arguments, allocator);
    // Blackholes program output.
    child.stdin_behavior = .Ignore;
    child.stdout_behavior = .Ignore;
    child.stderr_behavior = .Ignore;

    // If errored out, the program is *probably* not available.
    _ = child.spawnAndWait() catch return false;
    return true;
}

const PlayStrategy = *const fn (
    allocator: Allocator,
    song: Song,
) PlayStrategyError!void;

const PlayStrategyError = Child.SpawnError;

fn mpvPlayStrategy(allocator: Allocator, song: Song) PlayStrategyError!void {
    switch (song.format) {
        .flac, .mp3, .vorbis, .wav => {
            const arguments = [_][]const u8{
                "mpv",
                "--no-audio-display", // Prevents display of cover art.
                song.file_path,
            };

            var child = Child.init(&arguments, allocator);
            // TODO: handle exit code.
            _ = try child.spawnAndWait();
        },
    }
}

fn cvlcPlayStrategy(allocator: Allocator, song: Song) PlayStrategyError!void {
    switch (song.format) {
        .flac, .mp3, .vorbis, .wav => {
            const arguments = [_][]const u8{
                "cvlc",
                "--play-and-exit", // Makes exit after the song ends.
                song.file_path,
            };

            var child = Child.init(&arguments, allocator);
            // TODO: handle exit code.
            _ = try child.spawnAndWait();
        },
    }
}
